src/librustc/hir/lowering.rs

```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_generic_arg(&mut self,
-                         arg: &ast::GenericArg,
-                         itctx: ImplTraitContext<'_>)
-                         -> hir::GenericArg {
+    fn lower_generic_arg(
+        &mut self,
+        arg: &ast::GenericArg,
+        itctx: ImplTraitContext<'_>
+    ) -> hir::GenericArg {
         match arg {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
```

Comment:
The condition should be `!partial_res.base_res().matches_ns(TypeNS)`, then `Res::Err` won't need a special treatment.

I'm pretty sure `CtorKind::Fictive` shouldn't be treated specially too (or should be treated specially in resolve at worst), but I haven't got to the resolve part yet.

Reply:
`!partial_res.base_res().matches_ns(TypeNS)` would also allow it if it had resolved in the macro namespace, correct? Perhaps `!partial_res.base_res().matches_ns(TypeNS) && partial_res.base_res().matches_ns(ValueNS)`?

Reply:
All inappropriate resolutions are mapped into `Res::Err` in resolve, so later stages widely assume that you can't get e.g. `Def::Fn` from a type path, we can rely on it here as well.